### PR TITLE
MultiVI user guide

### LINKED ISSUE #1173: User guide: MultiVI

### PR DESCRIPTION
**Pull request by marianogabitto**
- Fixes #1173

### CODE REVIEW

File: docs/user_guide/models/multivi.rst [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+estimates the probability of accessibility.
+
+This generative process is also summarized in the following graphical model:
+
+.. figure:: figures/multivi_annotated_graphical_model.png
+   :class: img-fluid
+   :align: center
+   :alt: scVI graphical model
+
+   MultiVI graphical model.
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:98]:
```suggestion
```

File: docs/user_guide/models/multivi.rst [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+   * - :math:`\theta_g \in (0, \infty)`
+     - Inverse dispersion for negative binomial. This can be set to be gene/batch specific for example (and would thus be :math:`\theta_{kg}`), by passing ``dispersion="gene-batch"`` during model intialization. Note that ``px_r`` also refers to the underlying real-valued torch parameter that is then exponentiated on every forward pass of the model.
+   * - :math:`p_r`
+     - Acessibility probability estimate
+   * - :math:`\ell_n \in \left[0,1\right]`
+     - Cell-wise scaling factor
+     - ``d``
+   * - :math:`r_j \in \left[0,1\right]`
+     - Region-wise scaling factor
+     - ``f`
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:127]:
```suggestion
     - ``f``
```

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: docs/user_guide/models/multivi.rst [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+For RNA-seq, the prior parameters :math:`\ell_\mu` and :math:`\ell_{\sigma^2}` are computed per batch as the mean and
+variance of the log library size over cells. The expression data are generated from a count-based likelihood
+distribution, which here, we denote as the :math:`\mathrm{ObservationModel}`. While by default the
+:math:`\mathrm{ObservationModel}` is a :math:`\mathrm{ZeroInflatedNegativeBinomial}` (ZINB) distribution parameterized
+by its mean, inverse dispersion, and non-zero-inflation probability, respectively, users can pass
+``gene_likelihood = "negative_binomial"`` to :class:`~scvi.model.MultiVI`, for example, to use a simpler
+:math:`\mathrm{NegativeBinomial}` distribution. For ATAC-seq, detecting a region as accessible (:math:`y_{nj} > 0`) is
+generated by a Bernoulli random variable which depends on a cell-specific latent variable :math:`z^{acc}_n`, which captures
+biological heterogeneity, and two auxiliary scaling factors :math:`\ell_n, r_j`, which account for cell-specific and
+region-specific technical effects, respectively.
```
[the remark this answers is not in the document]
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:73]: Accepted.

File: docs/user_guide/models/multivi.rst [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+Differential accessibility
+--------------------------
+
+Differential accessibility analysis is achieved with :func:`~scvi.model.MULTIVI.differential_accessibility`. MultiVI tests differences in accessibility of :math:`g_z\left( z_n, s_n \right)`.
+
+.. topic:: References:
+
+   .. [#ref1] Tal Ashuach\*, Mariano I. Gabitto\*, Michael I. Jordan, Nir Yosef (2021),
+        *MultiVI: deep generative model for the integration of multi-modal data*,
+        `Bioarxiv, under review `__.
```
[the remark this answers is not in the document]
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:231]: Accepted.

File: docs/user_guide/models/multivi.rst [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+MultiVI uses variational inference and specifically auto-encoding variational bayes
+(see :doc:`/user_guide/background/variational_inference`) to learn both the model parameters (the
+neural network params, dispersion params, etc.) and an approximate posterior distribution with the following factorization:
+
+ .. math::
+    :nowrap:
+
+    \begin{align}
+       q_\eta(z_n, \ell_n \mid x_n) :=
+       q_\eta(z_n \mid x_n, s_n)q_\eta(\ell_n \mid x_n).
```
[the remark this answers is not in the document]
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:136]: The variable that is part of the generative model is z_n, this is the variable that we are approximating. z_rna and z_acc  are part of the variational proposal. Given that the last two variables are gaussian, z_n is Gaussian. I have re-written and I hope that it meets your expectations :) .

File: docs/user_guide/models/multivi.rst [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+     - Description
+     - Code variable (if different)
+   * - :math:`z_n \in \mathbb{R}^d`
+     - Low-dimensional representation capturing the state of a cell.
+     - N/A
+   * - :math:`\rho_n \in \Delta^{G-1}`
+     - Denoised/normalized gene expression.
+     - ``px_scale``
+   * - :math:`\ell_n \in (0, \infty)`
+     - Library size for RNA. Here it is modeled as a latent variable, but the recent default for scVI is to treat library size as observed, equal to the total RNA UMI count of a cell. This can be controlled by passing ``use_observed_lib_size=False`` to :class:`~scvi.model.SCVI`.
```
[the remark this answers is not in the document]
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:111]: Not but we should implement it . I am removing it for now.

File: docs/user_guide/models/multivi.rst [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    :nowrap:
+
+    \begin{align}
+       q_\eta(z_n, \ell_n \mid x_n) :=
+       q_\eta(z_n \mid x_n, s_n)q_\eta(\ell_n \mid x_n).
+    \end{align}
+
+Here :math:`\eta` is a set of parameters corresponding to inference neural networks (encoders), which we do not describe in detail here,
+but are described in the MultiVI paper. The underlying class used as the encoder for MultiVI is :class:`~scvi.nn.Encoder`.
+In the case of ``use_observed_lib_size=True``, :math:`q_\eta(\ell_n \mid x_n)` can be written as a point mass on the observed library size.
```
[the remark this answers is not in the document]
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:141]: Agree, we should implement this in the future.

File: docs/user_guide/models/multivi.rst [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+.. math::
+     :nowrap:
+
+     \begin{align}
+        \mathbb{E}_{q_\eta(z_i \mid x_i)}\left[g_z\left( z_i, s_i \right) \right],
+     \end{align}
+
+The expectation is approximated using Monte Carlo, and the number of samples can be passed as an argument in the code::
+
+     >>> model.get_accessibility_estimates(n_samples=10)
```
[the remark this answers is not in the document]
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:206]: Yep, corrected. We are returning by default the mean of the encoder and then decoding p. 

Also, added how we use n_samples_overall

File: scvi/model/_multivi.py [hunk 9]
```diff
@@ -415,6 +415,12 @@ def get_accessibility_estimates(
         else:  # imputed is a list of tensors
             imputed = torch.cat(imputed).numpy()
 
+        # TO DO: This needs to be validates
+        if n_samples_overall > 1:
+            # The -2 axis correspond to cells.
+            imputed = np.concatenate(imputed, axis=-2)
+            imputed = imputed.mean(0)
+
```
Comment (adamgayoso) [inline scvi/model/_multivi.py:423]: let's not add this here

File: docs/user_guide/models/multivi.rst [hunk 10]
```diff
@@ -46,9 +46,7 @@ MultiVI posits that the observed UMI counts for cell :math:`n`, gene :math:`g`,
    :nowrap:
 
    \begin{align}
-    z^{acc}_n &\sim {\mathrm{Normal}}\left( {0,I} \right) \\
-    z^{rna}_n &\sim {\mathrm{Normal}}\left( {0,I} \right) \\
-    z_n &= \frac{\left( z^{rna}_n + z^{acc}_n \right)}{2} \\
+    z_n &= &\sim {\mathrm{Normal}}\left( {0,I} \right) \\
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:49]: I liked it better with the z acc and z rna and the averaging and then the removed paragraph before. It's still a bit amiguous when one modality is missing, you could introduce a mask?

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: docs/user_guide/models/multivi.rst (hunk 10 above)
Comment (marianogabitto) [reply docs/user_guide/models/multivi.rst:49]: I believe this is not accurate.

File: scvi/model/_multivi.py (hunk 9 above)
Comment (marianogabitto) [reply scvi/model/_multivi.py:423]: Ok

File: docs/user_guide/models/multivi.rst [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+    + Comprehensive in capabilities. Able to perform DE gene, DA region analysis.
+
+    + Scalable to very large datasets (>1 million cells).
+
+    + Once trained with sufficient multimodal data, able to accurately input missing modalities.
+
+The limitations of scVI include:
+
+    + Effectively requires a GPU for fast inference.
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:19]:
```suggestion
- Comprehensive in capabilities. Able to perform DE gene, DA region analysis.
- Scalable to very large datasets (>1 million cells).
- Once trained with sufficient multimodal data, able to accurately input missing modalities.

The limitations of MultiVI include:

- Effectively requires a GPU for fast inference.
```

File: docs/user_guide/models/multivi.rst (hunk 10 above)
Comment (adamgayoso) [reply docs/user_guide/models/multivi.rst:49]:
```suggestion
    z_n &\sim {\mathrm{Normal}}\left( {0,I} \right) \\
```

File: docs/user_guide/models/multivi.rst [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+(see :doc:`/user_guide/background/variational_inference`) to learn both the model parameters (the
+neural network params, dispersion params, etc.) and an approximate posterior distribution with the following factorization:
+
+ .. math::
+    :nowrap:
+
+    \begin{align}
+       q_\eta(z_n, \ell_n \mid x_n) :=
+       q_\eta(z_n \mid x_n, y_n, s_n)q_\eta(\ell_n \mid x_n).
+    \end{align}
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:130]:
```suggestion
.. math::
   :nowrap:

   \begin{align}
      q_\eta(z_n, \ell_n \mid x_n) :=
      q_\eta(z_n \mid x_n, y_n, s_n)q_\eta(\ell_n \mid x_n).
   \end{align}
```

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: docs/user_guide/models/multivi.rst [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+For RNA-seq, the prior parameters :math:`\ell_\mu` and :math:`\ell_{\sigma^2}` are computed per batch as the mean and
+variance of the log library size over cells. The expression data are generated from a count-based likelihood
+distribution, which here, we denote as the :math:`\mathrm{ObservationModel}`. While by default the
+:math:`\mathrm{ObservationModel}` is a :math:`\mathrm{ZeroInflatedNegativeBinomial}` (ZINB) distribution parameterized
+by its mean, inverse dispersion, and non-zero-inflation probability, respectively, users can pass
+``gene_likelihood = "negative_binomial"`` to :class:`~scvi.model.MultiVI`, for example, to use a simpler
+:math:`\mathrm{NegativeBinomial}` distribution.
+
+For ATAC-seq, detecting a region as accessible (:math:`y_{nj} > 0`) is
+generated by a Bernoulli random variable which depends on a cell-specific latent variable :math:`z^{acc}_n`, which captures
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:66]:
```suggestion
generated by a Bernoulli random variable which depends on a cell-specific latent variable :math:`z_n`, which captures
```

File: docs/user_guide/models/multivi.rst [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+   * - :math:`\theta_g \in (0, \infty)`
+     - Inverse dispersion for negative binomial. This can be set to be gene/batch specific for example (and would thus be :math:`\theta_{kg}`), by passing ``dispersion="gene-batch"`` during model intialization. Note that ``px_r`` also refers to the underlying real-valued torch parameter that is then exponentiated on every forward pass of the model.
+   * - :math:`p_r`
+     - Accessibility probability estimate
+   * - :math:`\ell_n \in \left[0,1\right]`
+     - Cell-wise scaling factor
+     - ``d``
+   * - :math:`r_j \in \left[0,1\right]`
+     - Region-wise scaling factor
+     - ``f``
```
Comment (adamgayoso) [inline docs/user_guide/models/multivi.rst:113]:
```suggestion
   * - :math:`\theta_g \in (0, \infty)`
     - Inverse dispersion for negative binomial. This can be set to be gene/batch specific for example (and would thus be :math:`\theta_{kg}`), by passing ``dispersion="gene-batch"`` during model intialization. Note that ``px_r`` also refers to the underlying real-valued torch parameter that is then exponentiated on every forward pass of the model.
     - N/A
   * - :math:`p_r`
     - Accessibility probability estimate
     - N/A
   * - :math:`\ell_n \in \left[0,1\right]`
     - Cell-wise scaling factor
     - ``d``
   * - :math:`r_j \in \left[0,1\right]`
     - Region-wise scaling factor
     - ``f``
```
